packages/astro/src/vite-plugin-astro-server/request.ts

```diff
@@ -27,10 +28,13 @@ export async function handleRequest(
 	// routing behavior matches production builds. This supports both file and directory
 	// build formats, and is necessary based on how the manifest tracks build targets.
 	const url = new URL(origin + req.url?.replace(/(index)?\.html$/, ''));
-	const pathname = decodeURI(url.pathname);
+	let pathname = decodeURI(url.pathname);
+	if(config.trailingSlash === 'never' && !req.url) {
+		pathname = '';
+	}
```

Comment:
nit: would be nice to skip `decodeURI` too if it would be assigned `''` depending on `config.trailingSlash === 'never' && !req.url`